### PR TITLE
Block things needing align8 from preinitializing

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
We don't align these correctly on the frozen heap.

Correctness first, perf second (it's questionable how important this is anyway).

### CODE REVIEW

**Review by jkotas — APPROVED** [review]
Thanks

**Review by filipnavara — APPROVED** [review]

**Comment by filipnavara** [comment]
This test failure seems new and relevant:
```
Unhandled exception. System.Exception: Exception of type 'System.Exception' was thrown.
at TestReferenceTypeAllocation.Run() + 0xa3
at Program.Main() + 0x7f
```

**Comment by MichalStrehovsky** [comment]
> Are `long/ulong` fields also marked as requiring `Align8`? If not, those are gonna fail on ARM32 with atomics too.

https://github.com/dotnet/runtime/blob/24a9ec6cf63ebac2e5e59b6dbdb8f1ff1d73fd6d/src/coreclr/tools/Common/TypeSystem/Common/TypeSystemHelpers.cs#L416-L445

File: src/tests/nativeaot/SmokeTests/Preinitialization/Preinitialization.cs [hunk 1]
```diff
@@ -404,7 +404,15 @@ public ReferenceType(int intValue, double doubleValue)
...
-        Assert.IsPreinitialized(typeof(TestReferenceTypeAllocation));
+        if (RuntimeInformation.ProcessArchitecture is Architecture.Arm or Architecture.Wasm)
+        {
+            // Because of the double field, this is not preinitialized
+            Assert.IsLazyInitialized(typeof(TestReferenceTypeAllocation));
+        }
+        else
+        {
+            Assert.IsPreinitialized(typeof(TestReferenceTypeAllocation));
+        }
```
Comment (MichalPetryka) [inline src/tests/nativeaot/SmokeTests/Preinitialization/Preinitialization.cs:415]: Could you add a test for `long/ulong` too?

Comment (MichalStrehovsky) [reply src/tests/nativeaot/SmokeTests/Preinitialization/Preinitialization.cs:415]: This is enough to exercise we go through the align8 paths. Testing the details of how align8 works is done elsewhere.

**Comment by filipnavara** [comment]
`Build linux-arm Release NativeAOT_Libs failed` failure is https://github.com/dotnet/runtime/issues/99079#issuecomment-1971393588.

**Pull request merged**